### PR TITLE
feat: Cache heavy NetworkInfo db-sync queries

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context

We need to cache the following values in the context of [DbSyncNetworkInfoProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/b70ae5222a54634ba50bba0c9cc73b03832afe00/packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts#L28-L31) to limit impact on the DB:

`this.#builder.queryTotalSupply(maxLovelaceSupply)` -  Constant for the epoch
`this.#builder.queryCirculatingSupply()` - Changes every block
`this.#builder.queryActiveStake()` - Constant for the epoch
`this.#builder.queryLiveStake()` -  Changes every block

The two constant values can be cached using knowledge of the current epoch, triggered at the point of epoch rollover, and then set a configurable value in minutes with a sensible default to cover the other two.

# Proposed Solution
**Capturing the epoch rollover ( cache invalidation for epoch values )**
- The implemented approach here is to just start up an interval on boot with a light query, and it run indefinitely.
- The optimized approach is to start a Chain-Sync Ogmios client with an epoch check in the `rollForward` handler, start polling the DB once we observe a rollover.

In order to know when there is epoch rollover, we query the latest one from `epoch` table and store it in the cache as current. The validation query is set to run every 10 seconds change, check if there is a new epoch with data created greater than the cached one, if so - override it and invalidate the cache.

**Validation requirements**
- Default caching duration of 120 minutes for queryCirculatingSupply  and queryLiveStake
- Duration must be an integer from 1 to 2880 (2 days). This does not impact queryTotalSupply or queryActiveStake, as the cache invalidation is defined by the epoch rollover.
- Polling interval is 10 seconds

# Important Changes Introduced
- add `DB_QUERIES_CACHE_TTL` env in `run` with custom range validation
- add `DB_POLL_INTERVAL` env in `run` 
- add `--db-queries-cache-ttl` option in `cli`with custom range validation
- add `--db-poll-interval` option in `cli`
- add [CacheService](https://github.com/input-output-hk/cardano-js-sdk/pull/276/files#diff-9e8c39a0834f189fcd5eb822e9a84eac51cbf6ab7b401e8fdfb160625a78044eR7) to cache heavy db queries (used `node-cache` package )
- add db sync polling on [NetworkInfoProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/12ce8b4c18d1df087fb5d1a7295ebe569aa6d188/packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts#L12) provider start (10 secs polling interval)
- create test utils [wrapWithTransaction](https://github.com/input-output-hk/cardano-js-sdk/blob/4f2694e56028ba76cf96bd4e93e9c7883d5850fd/packages/cardano-services/test/util.ts#L58) and [ingestDbData](https://github.com/input-output-hk/cardano-js-sdk/blob/4f2694e56028ba76cf96bd4e93e9c7883d5850fd/packages/cardano-services/test/util.ts#L40) helping where needs to be reproduced specific test cases but finally db snapshot and state shouldn't be mutated for other tests
- extend `Provider` and `HttpService` in order to keep services builder pure [(7ceca49):](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/7ceca494f698905e7e7492ad09bae5fbb250bea6)
-- extend `Provider` interface with `start` optional method
-- extend `DbSyncProvider` and `HttpSrvice` with `start` method
-- invoke services `start()` within `HttpServer.startImpl()`
-- remove db `healthCheck` during services creation, keep them pure
-- align tests and enhance `loadHttpServer.test.ts`

### CODE REVIEW

File: packages/cardano-services/src/CacheService/keys.ts
Comment (mkazlauskas) [inline packages/cardano-services/src/CacheService/keys.ts:1]: Maybe putting this under `/NetworkInfo` would be better. Also you can 'namespace' with an object, like `export const NetworkInfoCacheKey = {TOTAL_SUPPLY, ...}`

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services/test/Program/loadHttpServer.test.ts
Comment (IvayloAndonov) [inline packages/cardano-services/test/Program/loadHttpServer.test.ts:53]: This one is the problematic test where `getaddrinfo ENOTFOUND database.server.com` is thrown, caused by service [healthCheck](https://github.com/input-output-hk/cardano-js-sdk/blob/12ce8b4c18d1df087fb5d1a7295ebe569aa6d188/packages/cardano-services/src/NetworkInfo/NetworkInfoHttpService.ts#L34-L36). Same behavior if we add [ChainHistoryHttpService](https://github.com/input-output-hk/cardano-js-sdk/blob/64aa7aeff061aa2cf9bc6196347f6cf5b9c7f6be/packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts#L16) in `serviceNames` - the containerised db is not accesssible during the service creation when there is `db.query` health check prior inisialisation. Looks like a raise condition, this is the only failing test if all available services are added here.

Comment (mkazlauskas) [reply packages/cardano-services/test/Program/loadHttpServer.test.ts:53]: cc @iadmytro

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (IvayloAndonov) [inline packages/cardano-services/src/Program/loadHttpServer.ts:72]: This one could be passed as env var as well, any thoughts?

File: packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts]: Addressed [f0d1480](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/f0d148025fb2491bad63bde71a338d2be6f5869f)

File: packages/cardano-services/src/CacheService/keys.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/CacheService/keys.ts]: Addressed [e470d15](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/e470d156e91e9a5cbf074bf9a725d0b334256850)

File: packages/cardano-services/src/CacheService/CacheService.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/CacheService/CacheService.ts]: [e470d15](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/e470d156e91e9a5cbf074bf9a725d0b334256850)

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (IvayloAndonov) [inline packages/cardano-services/src/Program/loadHttpServer.ts:71]: `POLL_INTERVAL` could be passed as an env var as well, any thoughts?

Comment (mkazlauskas) [reply packages/cardano-services/src/Program/loadHttpServer.ts:71]: Agree, maybe `CACHE_POLL_INTERVAL`?

Comment (IvayloAndonov) [reply packages/cardano-services/src/Program/loadHttpServer.ts:71]: Actual the poll interval is not associated with the cache, rather than with the DB sync provider. The cache service shouldn't know about the provider's custom polling, no? 🤔

Comment (mkazlauskas) [reply packages/cardano-services/src/Program/loadHttpServer.ts:71]: Right. You don't have to use this exact name, just wanted to point out that POLL_INTERVAL is too generic for an env variable

Comment (rhyslbw) [reply packages/cardano-services/src/Program/loadHttpServer.ts:71]: `DB_POLL_INTERVAL` ?

File: packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts:87]:
```suggestion
      const DB_POLL_QUERIES_COUNT = 1;
```
1. We've been using this format for `number`, `bigint`, or `string` constants, although it would be good to have a lint rule to enforce it, so I'm ok with it either way.
2. We can just refer to this as DB polling since we're interfacing with PostgreSQL. `cardano-db-sync` is the projector.

File: packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts:104]:
```suggestion
        it('should query the DB only once when the response is cached', async () => {
```

File: packages/cardano-services/test/entrypoints.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:170]:
```suggestion
          expect.assertions(2);
          const cacheDurationOutOfRange = '3000';
```

File: packages/cardano-services/test/util.ts
Comment (rhyslbw) [inline packages/cardano-services/test/util.ts:55]:
```suggestion
 * Wraps integration test within a transaction and once test has finished, rollback is executed
```

Comment (IvayloAndonov) [reply packages/cardano-services/test/util.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/test/entrypoints.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts:171]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/NetworkInfo/NetworkInfoHttpService.test.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/src/CacheService/CacheService.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/CacheService/CacheService.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/src/NetworkInfo/NetworkInfoHttpService.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/NetworkInfo/NetworkInfoHttpService.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/Program/loadHttpServer.ts]: [6e5ee7c](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/6e5ee7c71babfe32d15e36f92ea25177a42f75a5)

File: packages/cardano-services/test/entrypoints.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/entrypoints.test.ts:258]:
```suggestion
          expect.assertions(2);
          const cacheDurationOutOfRange = '3000';
```

Sorry, I should have been more explicit. Just ensure any of the callback-style tests with assertions in handlers follow this pattern (eg. line 277 as well)

File: packages/cardano-services/test/entrypoints.test.ts
Comment (rhyslbw) [reply packages/cardano-services/test/entrypoints.test.ts:257]: Actually, this one is appropriate, it's the CLI that's wrong. Can you see if [this](https://github.com/input-output-hk/cardano-js-sdk/blob/b003d499b6fad289b8d9656c6293a4f23856b5ed/packages/cardano-services/src/cli.ts#L131) is being executed?

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cardano-services/src/InMemoryCache/InMemoryCache.ts
Comment (rhyslbw) [inline packages/cardano-services/src/InMemoryCache/InMemoryCache.ts:11]:
```suggestion
  constructor(cacheTtlInMins: number = CACHE_TTL_DEFAULT) {
```
Let's rename all variables for consistency: cache duration -> cache TTL

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cardano-services/test/entrypoints.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts:258]: Does it mean to add assertions in all other tests that are missing?

Comment (rhyslbw) [reply packages/cardano-services/test/entrypoints.test.ts:258]: Yes. It's to ensure future regression causing the assertions to not be executed do not produce a false-positive.

File: packages/cardano-services/test/entrypoints.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts:257]: This test is about `run.ts` only, not `cli` ( it exits with 0 properly ). It's failing [here](https://github.com/input-output-hk/cardano-js-sdk/blob/6e5ee7c71babfe32d15e36f92ea25177a42f75a5/packages/cardano-services/src/run.ts#L20-L22) where we do not have try-catch block.

File: packages/cardano-services/src/InMemoryCache/InMemoryCache.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/InMemoryCache/InMemoryCache.ts]: [00d4bba](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/00d4bbae9dd517b46acd8abc7125bcff480b513c)

File: packages/cardano-services/test/entrypoints.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/entrypoints.test.ts:265]: [00d4bba](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/00d4bbae9dd517b46acd8abc7125bcff480b513c)

File: packages/cardano-services/src/CacheService/CacheService.ts
Comment (IvayloAndonov) [reply packages/cardano-services/src/CacheService/CacheService.ts]: [00d4bba](https://github.com/input-output-hk/cardano-js-sdk/pull/276/commits/00d4bbae9dd517b46acd8abc7125bcff480b513c)

File: packages/cardano-services/.env.test
Comment (rhyslbw) [inline packages/cardano-services/.env.test:3]:
```suggestion
DB_QUERIES_CACHE_TTL=120
```

File: packages/cardano-services/src/Program/ProgramOptionDescriptions.ts
Comment (rhyslbw) [inline packages/cardano-services/src/Program/ProgramOptionDescriptions.ts:11]:
```suggestion
  DbQueriesCacheTtl = 'Db queries cache TTL in minutes between 1 and 2880',
```

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (rhyslbw) [inline packages/cardano-services/src/Program/loadHttpServer.ts:21]:
```suggestion
  dbQueriesCacheTtl: number;
```

File: packages/cardano-services/src/cli.ts
Comment (rhyslbw) [inline packages/cardano-services/src/cli.ts:28]:
```suggestion
import { cacheTtlValidator } from './util/validators';
```

File: packages/cardano-services/src/cli.ts
Comment (rhyslbw) [inline packages/cardano-services/src/cli.ts:81]:
```suggestion
    '--db-queries-cache-ttl <dbQueriesCacheTtl>',
```

File: packages/cardano-services/test/InMemoryCache/InMemoryCache.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/InMemoryCache/InMemoryCache.test.ts:7]: Best to just use the module name here
```suggestion
describe('InMemoryCache', () => {
```

File: packages/cardano-services/test/entrypoints.test.ts
Comment (rhyslbw) [reply packages/cardano-services/test/entrypoints.test.ts:263]: We just should have consistent behaviour between the run script and the CLI, so if `envalid` is doing the right thing, let's correct the CLI to produce the same exit code.

**Review by rhyslbw — CHANGES_REQUESTED** [review]
There are still some references to "duration", which I've highlighted some, but not all.

**Review by rhyslbw — APPROVED** [review]
Great work @IvayloAndonov. Just waiting for follow-up review from @mkazlauskas, then we're ready for squashing prior to QA review.

File: packages/cardano-services/src/CacheService/keys.ts
Comment (mkazlauskas) [reply packages/cardano-services/src/CacheService/keys.ts]: @IvayloAndonov can we move `packages/cardano-services/src/InMemoryCache/keys.ts` file to `packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/keys.ts`?

File: packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts
Comment (mkazlauskas) [inline packages/cardano-services/src/NetworkInfo/DbSyncNetworkInfoProvider/DbSyncNetworkInfoProvider.ts:70]: This would be better, wouldn't leak implementation details (setInterval):
```suggestion
    this.#pollService = pollDbSync(this.#cache, () => this.#builder.queryLatestEpoch(), this.#dbPollInterval);
    ...
    async close(): Promise<void> {
      this.#pollService?.shutdown();
      this.#cache.shutdown();
    }
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

**Review by mkazlauskas — DISMISSED** [review]
Nice work @IvayloAndonov! :tada:

File: packages/core/src/Provider/Provider.ts
Comment (mkazlauskas) [inline packages/core/src/Provider/Provider.ts:6]: Looking at this again, I'm actually not liking core Provider having `{start, close}`. Can we move those to cardano-services? I suggest `DbSyncProvider` or a new interface that extends core Provider. What do you think @rhyslbw?

**Comment by IvayloAndonov** [comment]
The commits are being squashed. Ready for QA :)

File: packages/core/src/Provider/Provider.ts
Comment (mkazlauskas) [reply packages/core/src/Provider/Provider.ts:6]: Makes perfect sense: `type RunnableProvider = Provider & RunnableModule`?

Comment (IvayloAndonov) [reply packages/core/src/Provider/Provider.ts:6]: I like the suggestion of `RunnableProvider`. So in the end we agree on: 
1. Leave only `healthCheck` method within `Provider` and get rid of [{start, close}](https://github.com/input-output-hk/cardano-js-sdk/blob/2c61c6eb2a7a7de81ce37c5d524e526bc17767de/packages/core/src/Provider/Provider.ts#L6-L13)?
2. Rename `RunnableModule`'s [start](https://github.com/input-output-hk/cardano-js-sdk/blob/71a939b874296d86183d89fce7877c565630e921/packages/cardano-services/src/RunnableModule.ts#L28-L32) to `run` and all inners methods (before/after run)?
3. Create `type RunnableProvider = Provider & Pick<RunnableModule, 'run' | 'shutdown'>` within `cardano-services` package
4. These two changes should be a separate commit as part of this PR?

Comment (mkazlauskas) [reply packages/core/src/Provider/Provider.ts:6]: Yes to all questions, unless @rhyslbw objects

Comment (rhyslbw) [reply packages/core/src/Provider/Provider.ts:6]: Just make sure to mark the method name change as breaking

Comment (mkazlauskas) [reply packages/core/src/Provider/Provider.ts:6]: I think `runAfter` setting running state is confusing. Perhaps not such a great refactor idea :thinking:

Comment (IvayloAndonov) [reply packages/core/src/Provider/Provider.ts:6]: Yeah, since we have `'starting' | 'running' | 'stopping'` states, I guess we shouldn't ignore `start` as a runnable module lifecycle method?

Comment (IvayloAndonov) [reply packages/core/src/Provider/Provider.ts:6]: It turned out there is another issue with the refactoring - as discussed we remove `{start/stop}` methods from `core`'s [Provider](https://github.com/input-output-hk/cardano-js-sdk/blob/a2bfcc62c25e71d78d07b961267d7ce9679b6cf4/packages/core/src/Provider/Provider.ts#L5) interface and create `type RunnableProvider<T extends Provider> = T & Pick<RunnableModule, 'start' | 'shutdown'>` inside `cardano-services` and start using it. But we have another packages such as [rabbitmqTxSubmitProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/3c0f6048c5cfa04654f0a5463dfccefd24c9054e/packages/rabbitmq/src/rabbitmqTxSubmitProvider.ts#L11) implementing [TxSubmitProvider](https://github.com/input-output-hk/cardano-js-sdk/blob/e155ed4efcd1338a54099d1a9034ccbeddeef1cc/packages/core/src/Provider/TxSubmitProvider/types.ts#L5) extending Provider. So the issue is we want to align all providers with `RunnableProvider` (located within `cardano-services`) where some of them depends only on the `core` package.
I'm happy to discuss the approach here in a quick huddle.

**Review by mkazlauskas — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]

**Comment by mkazlauskas** [comment]
Looks good to merge @rhyslbw ?

**Pull request merged**
